colossalai/pipeline/pipelinable.py

```diff
@@ -123,23 +142,10 @@ def to_layer_list(self, exec_seq=None):
         self._exec_seq = exec_seq
         if exec_seq is None:
             # if user do not provide the model executing sequence, we use the initialization order as the executing order.
-            children_name = []
+            
+            # flatten the model except children of ModuleList & Sequential
             for child in self._root_children:
-                layer_spec = self._layer_spec_dict[id(child)]
-                if layer_spec.typename in (torch.nn.modules.container.ModuleList,
-                                           torch.nn.modules.container.Sequential):
-                    for child_in_container in layer_spec.children:
-                        self._layer_spec_list.append(self._layer_spec_dict[id(child_in_container)])
-                        for name, module in self._model.named_modules():
-                            if id(module) == id(child_in_container):
-                                children_name.append(name)
-                                break
-                else:
-                    self._layer_spec_list.append(layer_spec)
-                    for name, module in self._model.named_modules():
-                        if id(module) == id(child):
-                            children_name.append(name)
-                            break
+                self._layer_spec_list = self._layer_spec_list + self._get_flatten_model(child)
```

Comment:
Why we have to trace the leaf submodule? For example, keep the GPTEmbedding is enough for us to do the partition, if users want to use leaf submodules of this Embedding module, such as col_nn.Embedding and dropout, they could use exec_seq to trace the different granularity submodules.

Reply:
I added this before I realized that the fwd is the key problem. This PR can be closed now. But recording a iterator in layerspec.children is still weird since even a len() method will cause troubles for children.

Reply:
I see.